src/main/kotlin/app/revanced/patcher/apk/Apk.kt

```diff
@@ -0,0 +1,353 @@
+             */
+            private val entries = module.listDexFiles().map {
+                BasicDexEntry(
+                    this,
+                    it.name,
+                    it.openStream().use { stream -> RawDexIO.readRawDexFile(stream, it.length, null) })
+            }.associateBy { it.entryName }
+
+            override fun getDexEntryNames() = entries.keys.toList()
+            override fun getEntry(entryName: String) = entries[entryName]
+        })
+        private val opcodes = dexFile.opcodes
+
+        /**
+         * The classes and proxied classes of the [Base] apk file.
+         */
+        val classes = ProxyBackedClassList(dexFile.classes)
+
+        /**
+         * Write [classes] to the archive.
+         *
+         */
+        internal fun writeDexFiles() {
+            // Make sure to replace all classes with their proxy.
+            val classes = classes.also(ProxyBackedClassList::applyProxies)
+            val opcodes = opcodes
+
+            // Create patched dex files.
+            mutableMapOf<String, MemoryDataStore>().also {
+                val newDexFile = object : org.jf.dexlib2.iface.DexFile {
+                    override fun getClasses() = classes.toSet()
+                    override fun getOpcodes() = opcodes
+                }
+
+                // Write modified dex files.
+                MultiDexIO.writeDexFile(
+                    true, -1, // Core count.
+                    it, Patcher.dexFileNamer, newDexFile, DexIO.DEFAULT_MAX_DEX_POOL_SIZE, null
+                )
+            }.forEach { (name, store) ->
+                module.apkArchive.add(object : InputSource(name) {
+                    override fun getMethod() = ZipEntry.DEFLATED
+                    override fun getLength(): Long = store.size.toLong()
+                    override fun openStream() = store.readAt(0)
+                })
+            }
+        }
+    }
+
+    /**
+     * Metadata about an [Apk] file.
+     *
+     * @param packageName The package name of the [Apk] file.
+     * @param packageVersion The package version of the [Apk] file.
+     */
+    data class PackageMetadata(val packageName: String, val packageVersion: String) {
+        internal constructor(manifestBlock: AndroidManifestBlock) : this(
+            manifestBlock.packageName ?: "unnamed split apk file", manifestBlock.versionName ?: "0.0.0"
+        )
+    }
+
+
+    /**
+     * @param out The [File] to write to.
+     */
+    fun save(out: File) {
+        module.writeApk(out)
+    }
+
+    /**
+     * The split apk file that is to be patched.
+     *
+     * @param path The path to the apk file.
+     * @see Apk
+     */
+    class Split(module: ApkModule) : Apk(module) {
+        enum class Type {
+            LANGUAGE, LIBRARY, ASSET;
+        }
+
+        private companion object {
+            val architectures = setOf("armeabi_v7a", "arm64_v8a", "x86", "x86_64")
```

Comment:
architectures are only relevant for the split apk of type Library. Should polymorphism be used instead of an enum to differentiate between the types of splits?

Reply:
Using polymorphism would indeed be an option. However, I believe the static method should be reintroduced if we do decide to use polymorphism, because that does not leave room for users to select the wrong class.

What do you think?

Reply:
How would the static method switch between the architectures?

Reply:
> How would the static method switch between the architectures?

I am saying the `when` statement [here](https://github.com/Axelen123/revanced-patcher/blob/insan/src/main/kotlin/app/revanced/patcher/apk/ApkBundle.kt) should maybe be extracted to a companion object function if we want to have a class for each split type.

Reply:
Done.